lib/rules/consistent-this.js

```diff
@@ -9,35 +9,54 @@
 //------------------------------------------------------------------------------
 
 module.exports = function(context) {
+    var alias, allowEmptyVar;
+
+    alias = context.options[0];
+    allowEmptyVar = context.options[1] === "allow-empty-var";
+
+    /**
+     * Checks that an assignment to an identifier only assigns 'this' to the
+     * appropriate alias.
+     * @param {ASTNode} node The assigning node.
+     * @param {String} name The name of the identifier.
+     * @param {ASTNode} value The expression to be assigned.
+     * @returns {void}
+     */
+    function checkAssignment(node, name, value) {
+        var isThis = value !== null && value.type === "ThisExpression";
+
+        if (name === alias) {
+            if (!isThis) {
+                context.report(
+                    node,
+                    "Designated 'this' alias '{{alias}}' is not assigned " +
+                        "to the current execution context.",
+                    { alias: name }
+                );
+            }
+        } else if (isThis) {
+            context.report(
+                node,
+                "Unexpected alias '{{alias}}' for 'this'.",
+                { alias: name }
+            );
+        }
+    }
 
     return {
 
-        "VariableDeclaration": function(node) {
-            var alias = context.options[0];
-
-            node.declarations.forEach(function(declaration) {
-                if (declaration.id.name === alias &&
-                        !(declaration.init && declaration.init.type === "ThisExpression")) {
-
-                    context.report(
-                        node,
-                        "Designated 'this' alias '{{alias}}' is not assigned " +
-                            "to the current execution context.",
-                        { alias: declaration.id.name }
-                    );
-                }
-
-                if (declaration.init &&
-                        declaration.init.type === "ThisExpression" &&
-                        declaration.id.name !== alias) {
-
-                    context.report(
-                        node,
-                        "Unexpected alias '{{alias}}' for 'this'.",
-                        { alias: declaration.id.name }
-                    );
+        "VariableDeclaration": function (node) {
+            node.declarations.forEach(function (declaration) {
+                if (!allowEmptyVar || declaration.init !== null) {
+                    checkAssignment(node, declaration.id.name, declaration.init);
                 }
             });
+        },
+
+        "AssignmentExpression": function(node) {
```

Comment:
Should we also check that the assignment appears in the same scope as the declaration of the variable? It might be easier to check that, if a `self` variable is declared and not assigned to, the assignment appears is a set of assignments directly following the declarations.